The autosave and revision endpoints for templates and template parts incorrectly process file-based templates

anton-vlasenko opened the pull request:
<!--
Hi there! Thanks for contributing to WordPress!

Pull Requests in this GitHub repository **must** be linked to a ticket in the WordPress Core Trac instance (https://core.trac.wordpress.org), and are only used for code review. **No pull requests will be merged on GitHub.**

See the WordPress Handbook page on using PRs for Code Review more information: https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/

If this is your first time contributing, you may also find reviewing these guides first to be helpful:
- FAQs for New Contributors: https://make.wordpress.org/core/handbook/tutorials/faq-for-new-contributors/
- Contributing with Code Guide: https://make.wordpress.org/core/handbook/contribute/
- WordPress Coding Standards: https://make.wordpress.org/core/handbook/best-practices/coding-standards/
- Inline Documentation Standards: https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/
- Browser Support Policies: https://make.wordpress.org/core/handbook/best-practices/browser-support/
- Proper spelling and grammar related best practices: https://make.wordpress.org/core/handbook/best-practices/spelling/
-->

Trac ticket: https://core.trac.wordpress.org/ticket/61970

---
**This Pull Request is for code review only. Please keep all other discussion in the Trac ticket. Do not merge this Pull Request. See [GitHub Pull Requests for Code Review](https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/) in the Core Handbook for more details.**

anton-vlasenko committed:
Implement get_itemst() for template-parts.

anton-vlasenko committed:
Create the wp_template_part type test post.

anton-vlasenko committed:
Add tests for template_parts.
Assert response code.

anton-vlasenko committed:
Update the "create" tests.

anton-vlasenko committed:
Add tests for wp_template_part.

anton-vlasenko committed:
Fix @covers tags.

anton-vlasenko committed:
Fix test_get_items_with_data method.

anton-vlasenko committed:
Use 2 slashes for backward-compatibility reasons.

anton-vlasenko committed:
Return bad request response code when a template is file-based.

anton-vlasenko committed:
Use constants.

anton-vlasenko committed:
Add more unit tests.

anton-vlasenko committed:
Better test method names.

anton-vlasenko committed:
Remove empty lines.

anton-vlasenko committed:
Add more tests.

anton-vlasenko committed:
Add more tests.

anton-vlasenko committed:
Add template parts to the revisions test class.

anton-vlasenko committed:
Add more tests.

anton-vlasenko committed:
Implement more tests.

anton-vlasenko committed:
Implement more tests.

anton-vlasenko committed:
Implement more tests.

anton-vlasenko committed:
Implement more tests.

anton-vlasenko committed:
Fix typos.

anton-vlasenko committed:
Fix CS.

anton-vlasenko committed:
Fix the test.

anton-vlasenko committed:
Fix the @ticket numbers.

anton-vlasenko committed:
Implement more tests.

# Conflicts:
#	tests/phpunit/tests/rest-api/wpRestTemplateRevisionsController.php

anton-vlasenko committed:
Implement more tests.

# Conflicts:
#	tests/phpunit/tests/rest-api/wpRestTemplateRevisionsController.php

spacedmonkey reviewed (commented):
This is looking good so for. Needs testing.

anton-vlasenko committed:
Fix the Tests_REST_wpRestTemplateAutosavesController test.

anton-vlasenko committed:
Fix the Tests_REST_wpRestTemplateRevisionsController test.

anton-vlasenko commented:
> This is looking good so for. Needs testing.

Thank you for the review, @spacedmonkey.
I’ve added testing instructions and fixed the failing PHPUnit tests.

spacedmonkey commented on src/wp-includes/rest-api/endpoints/class-wp-rest-template-revisions-controller.php:
```diff
@@ -170,7 +170,17 @@ protected function get_parent( $parent_template_id ) {
+			);
+		}
+
+		$parent_post_id = isset( $template->wp_id ) ? $template->wp_id : 0;
+
+		if ( (int) $parent_post_id <= 0 ) {
```
```suggestion
		$parent_post_id = isset( $template->wp_id ) ? (int) $template->wp_id : 0;

		if ( $parent_post_id <= 0 ) {
```
Nitpic.

anton-vlasenko committed:
Move the cast operator inside the ternary expression for better readability.

Co-authored-by: Jonny Harris <spacedmonkey@users.noreply.github.com>

swissspidy committed:
Merge branch 'trunk' into fix/template-parts-autosaves-dont-work

swissspidy commented:
Committed in https://core.trac.wordpress.org/changeset/59605
